### PR TITLE
Improve publish gem tag management

### PR DESCRIPTION
**Pull request by cbaines**

### CODE REVIEW

**Comment by cbaines** [comment]
This should make publishing gems work ( @whoojemaflip and I are having issues with govuk_content_models and plek respectively), but the build will still fail as it will fail when pushTag attempts to update the non-existent release branch. If https://github.com/alphagov/govuk-puppet/pull/5601 works, then that will prevent pushTag from updating attempting to update the release branch.

**Comment by 36degrees** [comment]
It might just be my Friday afternoon brain, but I really can't get my head around the first commit message - I can't seem to parse it as a proper sentence.

**Comment by MatMoore** [comment]
I think we could lose the `return` statements inside the `taggedReleaseExists` and `versionAlreadyPublished` checks. Then if either ones fails it should be safe to rerun the job.

File: modules/govuk_jenkins/files/var/lib/jenkins/groovy_scripts/govuk_jenkinslib.groovy [hunk 1]
```diff
@@ -311,6 +298,19 @@ def publishGem(String repository, String branch) {
     sh("gem build ${repository}.gemspec")
     sh("gem push ${repository}-${version}.gem")
   }
+
+  def taggedReleaseExists = sh(
+    script: "git ls-remote --exit-status --tags origin v${version}",
```
Comment (36degrees) [inline modules/govuk_jenkins/files/var/lib/jenkins/groovy_scripts/govuk_jenkinslib.groovy]: Should this be `--exit-code` as per https://git-scm.com/docs/git-ls-remote.html? Doesn't work locally, at least on my Mac (Git 2.9.0)

**Comment by cbaines** [comment]
> It might just be my Friday afternoon brain, but I really can't get my head around the first commit message - I can't seem to parse it as a proper sentence.

I agree that its not great, would "Push version tag last in publishGem" be any better?

**Comment by 36degrees** [comment]
Sorry, it's the commit description I'm struggling with more than the summary…

> This means that issues with the tagging, either it failing and stopping the gem being published, or where the tag exists, but the gem is not published from stopping the gem being published.

File: modules/govuk_jenkins/files/var/lib/jenkins/groovy_scripts/govuk_jenkinslib.groovy (hunk 1 above)
Comment (cbaines) [inline modules/govuk_jenkins/files/var/lib/jenkins/groovy_scripts/govuk_jenkinslib.groovy]: Yep, good spot :)

**Comment by cbaines** [comment]
> Sorry, it's the commit description I'm struggling with more than the summary…

Ah, I've tried rewriting it, but I'm still struggling to make it clear. I'm happy to try again if you think it still needs some work.

**Comment by cbaines** [comment]
> I think we could lose the return statements inside the taggedReleaseExists and versionAlreadyPublished checks. Then if either ones fails it should be safe to rerun the job.

Ah yeah, that does make more sense. I've done this in 81d2dc9 .

**Comment by 36degrees** [comment]
New commit message is 👌 – thanks for changing that!

**Pull request merged**